Refactor/guess game logic

Renad-Alalfy opened the pull request:
- Refactor guess game crash when user is a guest
- Handle no internet connection 
- Handle total points for guest in the game 



https://github.com/user-attachments/assets/4d6ac523-5b2b-4f23-bf2d-c1d518fa7e80

Renad-Alalfy committed:
Refactor: Handle network errors and remove logging in guess game screens

Renad-Alalfy committed:
Refactor: Centralize HintUsageResult and update ViewModels

Renad-Alalfy committed:
Refactor: Add null check for userId in ResultViewModel

Renad-Alalfy committed:
Merge remote-tracking branch 'origin/develop' into refactor/guess-game-logic

Renad-Alalfy committed:
Refactor: Add try-catch block for account ID retrieval in ResultViewModel

Renad-Alalfy committed:
Refactor: Display 'not enough points' dialog for hint error

Renad-Alalfy committed:
Refactor: Simplify navigation and point update logic in ResultViewModel

Renad-Alalfy committed:
Refactor: Handle unauthenticated users in point system

Renad-Alalfy committed:
Merge branch 'develop' into refactor/guess-game-logic

Joseph-Sameh-0 committed:
Refactor: Optimize User Point Retrieval and Navigation

This commit introduces several key changes:

- **`GetUserPointUseCase` Refactor:**
    - Removed `getAccountIdUseCase` dependency.
    - `invoke` function now directly accepts `userId` as a parameter.
    - Simplified logic to directly return `gamePointsRepository.getUserGamePoints(userId)`.

- **ViewModel Updates for User Points:**
    - `GuessQuestionViewModel`, `GuessGameScreenViewModel`, and `ProfileViewModel` now fetch the `accountId` first and then pass it to `GetUserPointUseCase` to retrieve user points.
    - `GuessGameScreenViewModel` and `ProfileViewModel` store `accountId` in their respective UI states.

- **Navigation Enhancement:**
    - `GuessQuestionViewModel`, `GuessByImageViewModel`, and `ResultViewModel` now use `NavOptions.Builder().setPopUpTo(0, true).build()` when navigating to clear the back stack, ensuring a cleaner navigation flow after game completion or exiting.

- **Error Handling in `UserRepositoryImpl`:**
    - Modified `getAccountId()` to use `runCatching` for safer error handling, returning `null` if an exception occurs.

- **Minor Updates:**
    - `UpdatePointsUseCase`: Removed null check for `userId` as it's now guaranteed by upstream logic.
    - `GamePointsRepositoryImpl`: Streamlined `getUserGamePoints` mapping and error catching.
    - Removed an unused import (`LocalActivity`) in `GuessGameScreen.kt`.

Renad-Alalfy committed:
Fix: Implement retry functionality in GuessByImageViewModel

Renad-Alalfy committed:
Merge branch 'refactor/total-points-Id' into refactor/guess-game-logic

Renad-Alalfy committed:
Refactor: Initialize GuessQuestionUiState with navigation arguments in ViewModel

Renad-Alalfy committed:
Refactor: Pass Account ID to GetUserPointUseCase

Renad-Alalfy committed:
Update release notes

Joseph-Sameh-0 commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/screen/guessQuestionScreen/GuessQuestionViewModel.kt:
```diff
@@ -235,36 +265,32 @@ class GuessQuestionViewModel @Inject constructor(
                 updateState(
                     screenState.value.copy(
-                        questionText = updatedQuestion.content,
+                        questionText = updatedQuestion?.content!!,
```
don't use !! at all, use  ?: and default value

Joseph-Sameh-0 commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/screen/resultScreen/ResultViewModel.kt:
```diff
@@ -19,80 +20,56 @@ class ResultViewModel @Inject constructor(
-                    )
-                )
-            }
+    override fun onExitClicked() { //TODO: remove it
+        navigate(Destinations.Screen)
+    }
```
you don't use it it can be removed

Joseph-Sameh-0 commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/screen/resultScreen/ResultViewModel.kt:
```diff
@@ -19,80 +20,56 @@ class ResultViewModel @Inject constructor(
-                    )
-                )
-            }
+    override fun onBackToMenuClicked() { //TODO: remove it
+        navigate(Destinations.Screen)
+    }
```
you don't use it it can be removed

Renad-Alalfy committed:
Refactor: Simplify ResultScreen navigation and handle potential nulls in GuessQuestionViewModel

Renad-Alalfy committed:
Refactor: Remove redundant navigation tests in ResultViewModelTest

Renad-Alalfy committed:
remove unused functions in ResultScreen.kt

ahmedNaser7 commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/HintUsageResult.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.paris_2.domain.game.entity.Question
+
+ sealed class HintUsageResult {
```
why you use sealed class in domain , and how can represent an entity ?

ahmedNaser7 committed:
Refactor: Move GuessCard and GuessCardImageState to common components

The `GuessCard` composable and `GuessCardImageState` enum have been moved from the `designsystem` module to the `guessGameUi` module's common components.

This change also updates the `GuessByImageScreen` to construct the full image URL by prepending `GuessByImageViewModel.IMAGE_BASE_URL` to the image path.

Additionally, a dependency on the `safeimageviewer` project has been added to the `guessGameUi` module.

Joseph-Sameh-0 commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/screen/guessbyimage/GuessByImageViewModel.kt:
```diff
@@ -291,14 +289,7 @@ class GuessByImageViewModel @Inject constructor(
-
     companion object {
-        private const val HINT_COST = 10
+        const val IMAGE_BASE_URL =  "https://image.tmdb.org/t/p/w500"
```
why the view model in presentation layer knows the url?

mohammed-akkad commented on feature/guessGame/guessGameUi/src/main/java/com/feature/guessGame/guessGameUi/common/components/GuessGame.kt:
```diff
@@ -0,0 +1,283 @@
+                            Brush.verticalGradient(
+                                listOf(
+                                    Theme.colors.surface,
+                                    Color(0xFFD85895).copy(alpha = 0.12f),
```
why use hard core color here?
